File: java/reverse_words_in_a_string.java
public class Solution {
    void reverse(char[] s, int start, int end) {
        while (start<end) {
            char tmp=s[start];
            s[start]=s[end];
            s[end]=tmp;
            start++;
            end--;
        }
    }
    public String reverseWords(String s) {
        char[] str=s.toCharArray();
        reverse(str, 0, str.length-1);
        
        for (int i=0; i<str.length;) {
            if (str[i]!=' ') {
                int start=i;
                while (i<str.length && str[i]!=' ') {
                    i++;
                }
                reverse(str, start, i-1);
            } else {
                i++;
            }
        }
        
        int end=0;
        int i=0;
        while (i<str.length && str[i]==' ') i++;
        for (;i<str.length; i++) {
            if (str[i]!=' ' || (i>0 && str[i-1]!=' ')) {
                if (end!=i) {
                    str[end]=str[i];
                }
                end++;
            }
        }
        
        end--;
        while (end>=0 && str[end]==' ') end--;
        
        return new String(str, 0, end+1);
    }
}

File: cpp/reverse-words-in-a-string.cpp
151. Reverse Words in a String  QuestionEditorial Solution  My Submissions
Total Accepted: 112874 Total Submissions: 715351 Difficulty: Medium
Given an input string, reverse the string word by word.

For example,
Given s = "the sky is blue",
return "blue is sky the".

Update (2015-02-12):
For C programmers: Try to solve it in-place in O(1) space.

click to show clarification.
Clarification:
What constitutes a word?
A sequence of non-space characters constitutes a word.
Could the input string contain leading or trailing spaces?
Yes. However, your reversed string should not contain leading or trailing spaces.
How about multiple spaces between two words?
Reduce them to a single space in the reversed string.
Hide Company Tags Microsoft Snapchat Apple Bloomberg Yelp
Hide Tags String
Hide Similar Problems (M) Reverse Words in a String II

class Solution {
public:
    void reverseWords(string &s) {
        int n=s.size();
        reverse(s.begin(), s.end());
        for (int i=0; i<n; i++) {
            if (s[i]!=' ') {
                int j=i+1;
                while (j<n && s[j]!=' ') j++;
                reverse(s.begin()+i, s.begin()+j);
                i=j;
            }
        }
        
        int end=0;
        for (int i=0; i<n; i++) {
            if (s[i]!=' ' || (i>0 && s[i-1]!=' ')) {
                if (i!=end) {
                    s[end]=s[i];
                }
                end++;
            }
        }
        
        // remove last space if there is        
        s.resize(end);
        if (s.back()==' ') s.pop_back();
    }
};

